tasks/test-website.js

```diff
@@ -0,0 +1,163 @@
+
+            checkedHashes[link] = res;
+            return res;
+        });
+};
+
+var outputBrokenLinks = function (pageUrl, links) {
+    if (links.length) {
+        util.log('URL: ' + pageUrl);
+        util.log(util.colors.red('Broken links:'));
+
+        links.forEach(function (link) {
+            util.log(util.colors.red(link));
+        });
+
+        util.log();
+    }
+};
+
+// NOTE: Echoes the link if it is broken;
+// otherwise, resolves to null.
+var getBrokenLink = function (result) {
+    return new Promise(function (resolve) {
+        if (result.broken)
+            resolve(result.url.resolved || result.url.original);
+        else {
+            var link       = result.url.resolved;
+            var parsedLink = url.parse(link);
+
+            if (parsedLink.hash) {
+                resolve(Promise
+                    .resolve(checkedHashes[link] || isHashValid(link))
+                    .then(function (isValid) {
+                        return isValid ? null : link;
+                    }));
+            }
+            else
+                resolve(null);
+        }
+    });
+};
+
+var checkLinks = function () {
+    return new Promise(function (resolve) {
+        var linkTests = [];
+        var pageTests = [];
+
+        var siteChecker = new blc.SiteChecker({ excludeLinksToSamePage: false }, {
+            link: function (result) {
+                linkTests.push(getBrokenLink(result));
+            },
+            page: function (error, pageUrl) {
+                pageTests.push(Promise
+                    .all(linkTests)
+                    .then(function (res) {
+                        var brokenLinks = res.filter(function (value) {
+                            return !!value;
+                        });
+
+                        outputBrokenLinks(pageUrl, brokenLinks);
+                        return brokenLinks.length;
+                    }));
+            },
+            html: function (tree, robots, response, pageUrl) {
+                linkTests = [];
+                parsedDocs[pageUrl] = tree;
+            },
+            end: function () {
+                resolve(Promise
+                    .all(pageTests)
+                    .then(function (res) {
+                        return !!res.reduce(function (acc, value) {
+                            return acc + value;
+                        });
+                    }));
+            }
+        });
+
+        siteChecker.enqueue('http://localhost:8080/testcafe');
+    });
+};
+
+module.exports = function (websiteServer) {
+    return checkLinks()
+        .then(function (failed) {
+            return new Promise(function (resolve, reject) {
+                websiteServer.close(function () {
+                    if (failed)
+                        reject('Broken links found!');
+                    else
+                        resolve();
+                });
+            });
+        });
+};
```

Comment:
you can refactor this in the following way:
```js
module.exports = function (websiteServer) {
    function closeServer () {
        return new Promise(function (resolve) {
            websiteServer.close(resolve);
        });
    }

    var checkFailed = false;

    return checkLinks()
        .then(function (failed) {
            checkFailed = failed;
            return closeServer();
        })
        .then(function () {
            if (checkFailed)
                throw new Error('Broken links found!');
        });
};
```
With this you avoid deep nested callbacks